keyvault/certrefresher.go

```diff
@@ -0,0 +1,125 @@
+package keyvault
+
+import (
+	"context"
+	//nolint:gosec // sha1 only used to display cert thumbprint in logs for cross-verification with keyvault.
+	"crypto/sha1"
+	"crypto/tls"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+type EventualExpirationErr struct {
+	time.Time
+}
+
+func (e *EventualExpirationErr) Error() string {
+	return fmt.Sprintf("could not refresh before expiration on %v", e.Time.String())
+}
+
+type tlsCertFetcher interface {
+	GetLatestTLSCertificate(ctx context.Context, certName string) (tls.Certificate, error)
+}
+
+// CertRefresher offers a mechanism to present the latest version of a tls.Certificate from KeyVault, refreshed at an interval.
+type CertRefresher struct {
+	certName string
+	kvc      tlsCertFetcher
+	logger   *zap.Logger
+	m        sync.RWMutex
+	cert     *tls.Certificate
+}
+
+// NewCertRefresher returns a CertRefresher. When there's no error, the CertRefresher's GetCertificate method is ready
+// for use, returning a valid tls.Certificate fetched from KeyVault during construction.
+func NewCertRefresher(ctx context.Context, kvc tlsCertFetcher, l *zap.Logger, certName string) (*CertRefresher, error) {
+	cf := CertRefresher{
+		certName: certName,
+		kvc:      kvc,
+		logger:   l,
+	}
+
+	cert, err := cf.kvc.GetLatestTLSCertificate(ctx, cf.certName)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not fetch initial cert")
+	}
+
+	cf.cert = &cert
+	l.Info("initial certificate fetched", zap.Object("certificate", &cf))
+	return &cf, nil
+}
+
+// MarshalLogObject makes *CertRefresher a zapcore.ObjectMarshaler. Not concurrent-safe.
+func (c *CertRefresher) MarshalLogObject(encoder zapcore.ObjectEncoder) error {
+	encoder.AddString("cert name", c.certName)
+	encoder.AddString("sha1 thumbprint", sha1String(c.cert.Leaf.Raw))
+	encoder.AddString("expiration", c.cert.Leaf.NotAfter.String())
+	return nil
+}
+
+// GetCertificate returns the latest certificate fetched from KeyVault.
+func (c *CertRefresher) GetCertificate() *tls.Certificate {
+	c.m.RLock()
+	defer c.m.RUnlock()
+	return c.cert
+}
+
+// Refresh starts refreshing the certificate at the interval provided.
+// It blocks until context is done or refreshing fails.
+func (c *CertRefresher) Refresh(ctx context.Context, interval time.Duration) error {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Wrap(ctx.Err(), "refresh canceled")
+		case <-ticker.C:
+			nextTick := time.Now().Add(interval) // approx
+
+			// todo: add retries for transient errors
+			if err := c.refresh(ctx); err != nil {
+				// best effort to determine if we should stop the refresh loop
+				certExpires := c.cert.Leaf.NotAfter
+				if certExpires.Before(nextTick) {
+					return &EventualExpirationErr{certExpires}
```

Comment:
Instead of failing, would it make sense to use `ticker.Reset()` to bump the refresh rate so you'll catch it in time?